operator/controllers/instance_controller.go

```diff
@@ -0,0 +1,461 @@
+				Name:     "cluster-admin",
+			},
+		}
+		if err := ctrl.SetControllerReference(instance, desiredRoleBinding, r.Scheme); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		if err := r.Create(ctx, desiredRoleBinding); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+	}
+
+	currentPod := &corev1.Pod{}
+	if err := r.Get(ctx, namespacedName, currentPod); err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		desiredPod := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      podName,
+				Namespace: namespace,
+				// TODO: add proper labels.
+			},
+			Spec: corev1.PodSpec{
+				ServiceAccountName: podName,
+				RestartPolicy:      corev1.RestartPolicyNever,
+				Containers: []corev1.Container{{
+					Name:            "provisioning",
+					Image:           r.ProvisioningPodImage,
+					ImagePullPolicy: corev1.PullIfNotPresent,
+					Command:         []string{"/bin/bash"},
+					Args:            []string{"-c", provisioningScript},
+					Env: []corev1.EnvVar{
+						{Name: "NAME", Value: helmInstanceName},
+						{Name: "CHART", Value: plan.Spec.Provisioning.Chart.URL},
+						{Name: "NAMESPACE", Value: instance.Namespace},
+					},
+					VolumeMounts: []corev1.VolumeMount{{
+						Name:      "values",
+						ReadOnly:  true,
+						MountPath: "/etc/metabroker-provisioning/",
+					}},
+				}},
+				Volumes: []corev1.Volume{{
+					Name: "values",
+					VolumeSource: corev1.VolumeSource{
+						Secret: &corev1.SecretVolumeSource{SecretName: valuesSecretName},
+					},
+				}},
+			},
+		}
+		if err := ctrl.SetControllerReference(instance, desiredPod, r.Scheme); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		if err := r.Create(ctx, desiredPod); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if currentPod.Status.Phase == corev1.PodSucceeded {
+		if err := r.Delete(ctx, currentServiceAccount); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		if err := r.Delete(ctx, currentRoleBinding); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		if err := r.Delete(ctx, currentPod); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to process provisioning pod: %w", err)
+		}
+		// As soon as the pod gets completed successfully, delete it.
+		return ctrl.Result{}, nil
+	} else if currentPod.Status.Phase == corev1.PodFailed {
+		// TODO: what should we do when the pod fails?
+	}
+
+	return ctrl.Result{RequeueAfter: time.Second * 3}, nil
+}
+
+const provisioningScript = `#!/bin/bash
+
+set -o errexit -o nounset
+
+helm install "${NAME}" "${CHART}" \
```

Comment:
`exec` (so as to not leave a process behind, with the entailing issues about signal passing)?  The alternative is to just skip the shell part and use the commands directly, since it's just a single command anyway (which means we can pass in an array directly instead of making sure the variable substitution is working correctly).

Reply:
I added `catatonit` instead. 5956645